### PR TITLE
SubMesh and AttachMement

### PR DESCRIPTION
**Pull request by super626**
works on IOS and Android.

Not test on Windows and Linux.
1. SubMesh Support, 
   sprite3d->getSubMeshState(0)->setTexture(tex);
   sprite3d->getSubMeshState(0)->setVisible(true);
2. AttachMent, attach Sprite3D to another,
        auto sprite = Sprite3D::create("girl.c3b");
        auto weapon = Sprite3D::create("weapon.c3b");
        auto attachNode = sprite->getAttachNode("left hand");
        attachNode->addChild(weapon);
3. better resin support coming latter, together with fbx-conv

### CODE REVIEW

**Comment by super626** [comment]
Windows works now

**Comment by minggo** [comment]
Please test on supported platforms.

File: cocos/3d/CCAttachNode.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+{
+    static Mat4 mat;
+    mat.setIdentity();
+    auto parent = getParent();
+    if (parent)
+    {
+        mat = parent->getWorldToNodeTransform() * _attachBone->getWorldMat() * getNodeToParentTransform();
+    }
+    else
+        mat = _attachBone->getWorldMat() * getNodeToParentTransform();
```
Comment (minggo) [inline cocos/3d/CCAttachNode.cpp:66]: Please use `{}` for both `if` and `else` or not use them at all.

File: cocos/3d/CCAttachNode.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    
+    return mat;
+}
+
+void AttachNode::visit(Renderer *renderer, const Mat4& parentTransform, uint32_t parentFlags)
+{
+    Node::visit(renderer, parentTransform * _attachBone->getWorldMat(), Node::FLAGS_DIRTY_MASK);
+}
+
+NS_CC_END
```
Comment (minggo) [inline cocos/3d/CCAttachNode.cpp:75]: Please add a line at the end of a .cpp file.

File: cocos/3d/CCBundle3D.cpp [hunk 3]
```diff
@@ -350,13 +443,29 @@ bool Bundle3D::loadSkinDataJson(SkinData* skindata)
...
-    //skindata->nodeBoneOriginMatrices.resize(skindata->nodeBoneNames.size());
     getChildMap(skindata->boneChild, skindata, skin_data_1);
     return true;
 }
 
 bool Bundle3D::loadMaterialDataJson(MaterialData* materialdata)
 {
+    if (_version == "1.2")
+    {
+        return loadMaterialDataJson_0_1(materialdata);
```
Comment (minggo) [inline cocos/3d/CCBundle3D.cpp:456]: It is strange that `_version` is `1.2`, but load function is `0_1`.
And please don't use magic number.

Comment (minggo) [inline cocos/3d/CCBundle3D.cpp:456]: magic string.

File: cocos/3d/CCBundle3DData.h [hunk 4]
```diff
@@ -156,10 +157,58 @@ struct SkinData
...
+        rootBoneIndex = -1;
+    }
+    
+    void addBoneNames(const std::string& name)
+    {
+        for (auto iter : boneNames)
+        {
+            if ((iter) == name)
+                return;
+        }
```
Comment (minggo) [inline cocos/3d/CCBundle3DData.h]: Is it better to use `std::binary_search()`?

Comment (minggo) [inline cocos/3d/CCBundle3DData.h]: Oh, it is not a good idea because it is not sorted. 
Please ignore it.

Comment (minggo) [inline cocos/3d/CCBundle3DData.h]: But how about `std::find()`?

File: cocos/3d/CCMeshSkin.cpp [hunk 5]
```diff
@@ -390,10 +185,14 @@ Vec4* MeshSkin::getMatrixPalette()
...
+    int i = 0, paletteIndex = 0;
+    static Mat4 t;
 	for (auto it : _skinBones )
 	{
-        it->updateJointMatrix(&_matrixPalette[i++ * PALETTE_ROWS]);
+        Mat4::multiply(it->getWorldMat(), _invBindPoses[i++], &t);
+        _matrixPalette[paletteIndex++].set(t.m[0], t.m[4], t.m[8], t.m[12]);
+        _matrixPalette[paletteIndex++].set(t.m[1], t.m[5], t.m[9], t.m[13]);
+        _matrixPalette[paletteIndex++].set(t.m[2], t.m[6], t.m[10], t.m[14]);
 	}
```
Comment (minggo) [inline cocos/3d/CCMeshSkin.cpp]: Wrong indent.

File: cocos/3d/CCSkeleton3D.cpp [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+Bone3D* Bone3D::getParentBone()
+{
+    return _parent;
+}
+ssize_t Bone3D::getChildBoneCount() const
+{
+    return _children.size();
+}
+Bone3D* Bone3D::getChildBoneByIndex(int index)
```
Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp]: getChildBoneByIndex(int index) -> getChildBoneByIndex(int index) const

Please check other functions and add const as possible.

File: cocos/3d/CCSkeleton3D.cpp [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+
+void Bone3D::updateLocalMat()
+{
+    if (_blendStates.size())
+    {
+        Vec3 translate(Vec3::ZERO), scale(Vec3::ZERO);
+        Quaternion quat(Quaternion::zero());
+        
+        float total = 0.f;
+        for (auto it: _blendStates) {
```
Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp:208]: Please add `{` in a new line as other codes did.

File: cocos/3d/CCSkeleton3D.cpp [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+{
+    removeAllChildBone();
+}
+
+void Bone3D::updateLocalMat()
+{
+    if (_blendStates.size())
+    {
+        Vec3 translate(Vec3::ZERO), scale(Vec3::ZERO);
+        Quaternion quat(Quaternion::zero());
```
Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp]: Can we also have `Quaternion ::ZERO`?

File: cocos/3d/CCSkeleton3D.cpp [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+{
+    removeAllBones();
+}
+
+//create a new meshskin if do not want to share meshskin
+Skeleton3D* Skeleton3D::create(const std::string& filename, const std::string& name)
+{
+    //load skin here;
+    std::string fullPath = FileUtils::getInstance()->fullPathForFilename(filename);
+    std::string key = fullPath + "#" + name;
```
Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp]: Why not use `std::string key(fullPath + "#" + name)`?

File: cocos/3d/CCSkeleton3D.cpp [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+        for (auto childIt : it.second) {
+            auto child = getBoneByIndex(childIt);
+            child->_parent = parent;
+            parent->_children.pushBack(child);
+        }
+    }
+    
+    setRootBone(getBoneByIndex(skeletondata.rootBoneIndex));
+    if (_rootBone)
+    _rootBone->resetPose();
```
Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp]: Please add 4 spaces for `_rootBone->resetPose()`.

**Comment by super626** [comment]
I have tested all platforms except linux. We do not have a linux machine. Can you or Harrison help?

File: cocos/3d/CCBundle3DData.h (hunk 4 above)
Comment (super626) [inline cocos/3d/CCBundle3DData.h]: Yes, it is not sorted. Use std::find can make the code more clear. We will use it.

File: cocos/3d/CCAttachNode.cpp (hunk 1 above)
Comment (super626) [inline cocos/3d/CCAttachNode.cpp:66]: got it

File: cocos/3d/CCAttachNode.cpp (hunk 2 above)
Comment (super626) [inline cocos/3d/CCAttachNode.cpp:75]: got it

**Comment by minggo** [comment]
Please set up all supported environments. 
We can help to test it this time, but we can not do it all the time.
And please don't forget to set up WP8 and win8 environments.

File: cocos/3d/CCBundle3D.cpp (hunk 3 above)
Comment (super626) [inline cocos/3d/CCBundle3D.cpp:456]: we will modify it.

Comment (super626) [inline cocos/3d/CCBundle3D.cpp:456]: the 1.2 is a wrong version. We use 1.2 for the first version. So can we add comment to this and use 0_1?

Comment (minggo) [inline cocos/3d/CCBundle3D.cpp:456]: No problem.

**Comment by minggo** [comment]
And please fix conflict.
It is easy to have conflict because modifying project file.

**Comment by minggo** [comment]
I will merge it ASAP.

File: cocos/3d/CCSkeleton3D.cpp (hunk 7 above)
Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp:208]: It is not problem to use this style.
But it is a better to use the same style in a file.
It is not a big deal.

File: cocos/3d/CCSkeleton3D.cpp (hunk 8 above)
Comment (super626) [inline cocos/3d/CCSkeleton3D.cpp]: Quternion does not have ZERO

Comment (minggo) [inline cocos/3d/CCSkeleton3D.cpp]: I see.
Why not add it?
Just a suggestion.

Comment (super626) [inline cocos/3d/CCSkeleton3D.cpp]: Yes, it is better to add it. I will do it

**Comment by minggo** [comment]
And please fix compiling error.
It seems it has problem in lua-binding.

**Comment by minggo** [comment]
[ci rebuild]

**Comment by minggo** [comment]
[ci rebuild]

**Pull request merged**